engine/src/main/java/org/terasology/rendering/dag/RenderGraph.java

```diff
@@ -15,44 +15,138 @@
+
+        if (nodeMap.containsKey(nodeUri)) {
+            int i = 2;
+            while (nodeMap.containsKey(new SimpleUri(nodeUri.toString() + i))) {
+                i++;
+            }
+            nodeUri = new SimpleUri(nodeUri.toString() + i);
+        }
+
+        nodeMap.put(nodeUri, node);
         node.setUri(nodeUri);
+
         return nodeUri;
     }
 
+    public Node removeNode(SimpleUri nodeUri) {
+        Validate.notNull(nodeUri, "nodeUri cannot be null!");
+
+        if (edgeMap.containsKey(nodeUri)) {
+            throw new RuntimeException("The node you are trying to remove is still connected to other nodes in the graph!");
+        }
+
+        return nodeMap.remove(nodeUri);
+    }
+
     public Node findNode(SimpleUri nodeUri) {
-        for (Node node: nodes) {
-            if (node.getUri().equals(nodeUri)) {
-                return node;
-            }
+        Validate.notNull(nodeUri, "nodeUri cannot be null!");
+
+        return nodeMap.get(nodeUri);
+    }
+
+    public boolean connect(Node fromNode, Node toNode) {
+        Validate.notNull(fromNode, "fromNode cannot be null!");
+        Validate.notNull(toNode, "toNode cannot be null!");
+
+        reverseEdgeMap.put(toNode, fromNode);
+        boolean success = edgeMap.put(fromNode, toNode);
+        if (!success) {
+            logger.warn("Trying to connect two already connected nodes, " + fromNode.getClass() + " and " + toNode.getClass());
         }
 
-        return null;
+        return success;
+    }
+
+    public boolean disconnect(Node fromNode, Node toNode) {
+        Validate.notNull(fromNode, "fromNode cannot be null!");
+        Validate.notNull(toNode, "toNode cannot be null!");
+
+        reverseEdgeMap.remove(toNode, fromNode);
+        return edgeMap.remove(fromNode, toNode);
     }
 
-    // TODO: add remove, get, addEdge, removeEdge methods here
+    // TODO: Add `boolean isFullyFunctional(Node node)`
+
+    // TODO: Add handler methods which the graph uses to communicate changes to a node.
 
     public List<Node> getNodesInTopologicalOrder() {
-        return nodes;
+        List<Node> topologicalList = new ArrayList<>();
+
+        Map<Node, Integer> inDegreeMap = Maps.newHashMap();
+        List<Node> nodesToExamine = Lists.newArrayList();
+        int visitedNodes = 0;
+
+        // Calculate the in-degree for each node.
+        for (Entry<Node, Collection<Node>> nodeEdgeList : reverseEdgeMap.asMap().entrySet()) {
+            inDegreeMap.put(nodeEdgeList.getKey(), nodeEdgeList.getValue().size());
+        }
+
+        // Add the missing nodes that did not have any edges, and mark them for examination
+        for (Node node : nodeMap.values()) {
+            if (!inDegreeMap.containsKey(node)) {
+                inDegreeMap.put(node, 0);
+                nodesToExamine.add(node);
+            }
+        }
+
+        while (!nodesToExamine.isEmpty()) {
+            Node currentNode = nodesToExamine.remove(0);
```

Comment:
I'm pretty sure this is perfectly safe. As you mentioned, the problem arises only when you modify the collection while ITERATING over it. The reason is usually indices:
```
for (int i = 0; i < list.size(); i++) {
     if (something) {
         list.removeAt(i);  <-- This effectively changes the address of all succeeding elements by 1. One workaround is to subtract i by 1 here, but it's usually not used.
    }
}
```
As demonstrated in the example, the problem is because of indices. I'm not using indices at all - for that matter, I'm not "iterating" the collection at all. I'm just checking if it's empty, and popping if it's not empty.

That being said, I'll look up some more to make sure.

Reply:
Okay, I checked quite a few places - this kind of access is used a lot, and seems to be perfectly safe.

Reply:
Aaaaalright then.